presto-main/src/main/java/com/facebook/presto/sql/planner/GroupingOperationRewriter.java

```diff
@@ -77,48 +82,100 @@ public static Expression rewriteGroupingOperation(GroupingOperation expression,
+            resolveFunction(rewritten, functionArgumentTypes, functionRegistry);
 
             return rewritten;
         }
     }
 
-    private static Expression translateFieldToLongLiteral(FieldId fieldId, RelationId requiredOriginRelationId)
+    private static int translateFieldToInteger(FieldId fieldId, RelationId requiredOriginRelationId)
     {
         // TODO: this section should be rewritten when support is added for GROUP BY columns to reference an outer scope
         checkState(fieldId.getRelationId().equals(requiredOriginRelationId), "grouping arguments must all come from the same relation");
-        return new LongLiteral(Integer.toString(fieldId.getFieldIndex()));
+        return fieldId.getFieldIndex();
+    }
+
+    /**
+     * The grouping function is used in conjunction with GROUPING SETS, ROLLUP and CUBE to
+     * indicate which columns are present in that grouping.
+     *
+     * <p>The grouping function must be invoked with arguments that exactly match the columns
+     * referenced in the corresponding GROUPING SET, ROLLUP or CUBE clause at the associated
+     * query level. Those column arguments are not evaluated and instead the function is
+     * re-written with the arguments below.
+     *
+     * <p>To compute the resulting bit set for a particular row, bits are assigned to the
+     * argument columns with the rightmost column being the most significant bit. For a
+     * given grouping, a bit is set to 0 if the corresponding column is included in the
+     * grouping and 1 otherwise. For an example, see the SQL documentation for the
+     * function.
+     *
+     * @param groupId An ordinal indicating which grouping is currently being processed.
+     *        Each grouping is assigned a unique monotonically increasing integer.
+     * @param columns The column arguments with which the function was
+     *        invoked converted to ordinals with respect to the base table column
+     *        ordering.
+     * @param groupingSetDescriptors A collection of ordinal lists where the index of
+     *        the list is the groupId and the list itself contains the ordinals of the
+     *        columns present in the grouping. For example: [[0, 2], [2], [0, 1, 2]]
+     *        means the the 0th list contains the set of columns that are present in
+     *        the 0th grouping.
+     * @return A bit set converted to decimal indicating which columns are present in
+     *         the grouping. If a column is NOT present in the grouping its corresponding
+     *         bit is set to 1 and to 0 if the column is present in the grouping.
+     */
+    static long calculateGrouping(long groupId, List<Integer> columns, List<List<Integer>> groupingSetDescriptors)
+    {
+        long grouping = (1L << columns.size()) - 1;
+
+        List<Integer> groupingSet = groupingSetDescriptors.get((int) groupId);
+        for (Integer groupingColumn : groupingSet) {
+            int index = columns.indexOf(groupingColumn);
```

Comment:
I know this is just copied over from previous place, but usage of `indexOf` is slightly surprising. What about:

```
Set groupingSet = new HashSet(groupingSetDescriptors.get(groupId));
for (int index = 0; index < columns.size(); index++) {
    if (groupingSet.contains(columns.get(index))) {
      ...
    }
}
```

Reply:
What you suggest will work. However, the `columns` list will never have more than 64 elements making the `indexOf()` operation essentially constant. It's unclear if the overhead of building a `HashSet` for every `groupingSet` will be worth it. It certainly will be worth it when the `groupingSet` is large and the `columns` list is small. However, the current code will probably be more efficient when the `groupingSet` is small and the the `columns` list is large. Anyway, my point is that it's not obvious that what you suggest is strictly better, unless I'm missing something. I think the current code and what you suggest will perform equally.

Reply:
Yea... For small data it surely doesn't matter. So you can use `Map` an be on the safe side (considering also future modifications to this code by authors less aware of currently applicable data size considerations)

Reply:
I'm sorry I'm not following. Do you mean `Set` instead of `Map`? I still don't see how what you suggest is strictly better.